MS Intune Mobile App Support config, setup, troubleshooting, and end user workflows (#8599)
mattermost/docs: source/deployment-guide/mobile/configure-microsoft-intune-mam.rst

If any of the following apply, stop. This configuration will fail.

* You require Android Intune MAM support (not yet available).
* Your deployment cannot use Microsoft Entra ID (Azure AD).
* Your identity strategy cannot use Azure AD ``objectId`` as the authoritative user identifier.
* You need a rollout model where users can defer or bypass Intune enrollment.

Before You Continue
-------------------

Before proceeding, confirm the following are true:

* You use Microsoft Entra ID for authentication.
* You can commit to Azure AD ``objectId`` as the authoritative identity.
* You have (or can obtain) a Mattermost Enterprise Advanced license.
* Target users are licensed for Microsoft Intune.
* You can register applications and grant admin consent in Microsoft Entra.

If any of the above are not true, do not proceed.

Configuration Overview
----------------------

Configuring Intune MAM for Mattermost Mobile requires coordinated setup across the following four systems:

* **Microsoft Entra ID (Azure AD)** – identity, app registration, API permissions
* **Microsoft Intune** – app protection policies and user targeting
* **Mattermost Server** – MAM enablement and identity alignment

> * **Mattermost Mobile App (iOS)** – enrollment and enforcement

Reviewer: there is nothing to configure on the mobile app

Author: This section is intended to give admins a broad overview of the work ahead across multiple systems. This bulleted list is introduced "coordinated setup across the following systems", and the mobile app setup is enrollment and enforcement.

Let me know if you'd like changes made to this area.

Reviewer: I guess this is fine, just want to provide clarity that there is nothing that needs to be configured in the mobile app itself.

If any system is misconfigured, Intune MAM enrollment will fail.

Before beginning configuration, review the `Identity Configuration <#identity-configuration-for-intune-mam>`__ section to confirm your deployment meets the required identity model.

Setup Order
~~~~~~~~~~~

Follow this setup order exactly to avoid enrollment failures and rework.

Step 1: Finalize Identity Strategy (Entra + Mattermost)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* Commit to Azure AD ``objectId`` as the authoritative identity.

> * Ensure all authentication methods (OAuth, SAML, LDAP) resolve to the same value.

Reviewer: ```suggestion
* Ensure all authentication methods (OpenID Connect or SAML, LDAP) resolve to the same value.
```

OpenID Connect will always be `objectId` while SAML and LDAP need configured to use the value of `objectId`, at the same time SAML is only required to be that way if it is the selected `Auth Provider` in the Intune MAM configuration in the admin console.

> * Confirm access tokens include the ``oid`` claim.

Reviewer: this will be part of the configuration steps when configuring in Entra, not sure this is needed here.

Reviewer: I mean, this is something that they will HAVE to configure in Entra when they register the application (This is not registering the mobile application, just an Entra application, then during the configuration of that Application they will need to add some claims the the access token, but the access token even without optional claims, will include the `oid`

Screenshot for reference

<img width="2075" height="485" alt="Image" src="https://github.com/user-attachments/assets/4bbf2aa8-e10a-457d-a02f-15b97990f5b9" />

Identity must be finalized before enabling Intune or enrolling users.

Step 2: Configure Microsoft Entra Applications & Permissions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* Register the Mattermost mobile app as a public/native Entra application.
* Copy the **Application (Client) ID**.

  To register an application in Microsoft Entra, see: https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app

* Configure the iOS platform with the correct bundle ID and redirect URI.

  For Entra portal steps, see: https://learn.microsoft.com/en-us/entra/identity-platform/scenario-mobile-app-configuration

  For redirect URI formatting details, see: https://learn.microsoft.com/en-us/entra/identity-platform/how-to-add-redirect-uri

* Grant required Intune MAM API permissions with admin consent.

>   To grant tenant-wide admin consent, see: https://learn.microsoft.com/en-us/entra/identity/enterprise-apps/grant-admin-consent

Reviewer: not at all.

Register an Application in Entra that will be use to authenticate the users.

this link works `To register an application in Microsoft Entra, see: https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app` and we can keep that

The redirect URI should remain empty

Once the app is registered, they should `Expose an API` more info here: https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-configure-app-expose-web-apis

The API to be exposed should be `api://<APPLICATION-ID>` normally filled automatically, then add a scope named `login.mattermost`

Once that is done, they need to `Add a client application`, add the Client Id as listed below and assign the `Authorized scopes` to `api://<APPLICATION-ID>/login.mattermost`
- Mattermost Mobile Beta: `64e9952b-20eb-46dc-92ad-99089ed24903`
- Mattermost Mobile: `not yet created, we will need to update this document`

in addition I will share a script once we have the Client Id for `Mattermost Mobile` so we can attach it here.

Then in `Token configuration` -> `Add optional claim`, Select Token type `Access` then select the claims
- email
- family_name
- given_name
- preferred_username
- upn

Author: To finalize the documentation, I'll need:

- Production Mobile Client ID
- Confirmation that ``login.mattermost`` is the final scope name
- Confirmation that all listed access token claims are mandatory
- The script you mentioned (or guidance on where it will live)

Step 3: Configure Mattermost Server for Intune MAM
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* Enable Intune MAM in the System Console.
* Set ``IdAttribute = objectId``.

> * Verify Enterprise Advanced licensing.

Reviewer: Here they will need to add the `Application (client) ID` and the `Directory (tenant) ID` of the Application they just registered in Step 2 as well as Selecting their Auth Provider (available based on what is configured OpenID Connect with Office 365 or SAML).

Important: For SAML we are not really validating that the IdP is Entra, so perhaps we must make absolutely sure they understand that to use SAML they must use it with Entra ID (I know is in Before you Continue, but we can mention it here too)

Step 4: Configure Intune App Protection Policies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* Create an iOS App Protection Policy.

> * Add the Mattermost bundle ID.

Reviewer: bundleIds

- Mattermost Mobile Beta: `com.mattermost.rnbeta`
- Mattermost Mobile: `com.mattermost.rn`

* Assign the policy using Azure AD groups.

Step 5: Validate Using the Mobile App
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> * Ensure test users are assigned in Intune and properly licensed.

Reviewer: Here, we need to add a troubleshooting section in case they face that the user did not consent permissions, I'll explain shortly.

Also very important that they do this with an account that has capacity to consent to permissions for the entire organization, and once they are presented with the consent form during the login flow, the should mark the option to consent to the permissions on behalf of the org.

Troubleshooting: "**Consent required**"
There can be a situation when after logging in for the first time, the enrollment does not happen and consent triggers as being required.

To solve this (as an Entra admin), go to Enterprise Apps, search for (Mattermost Mobile Beta or Mattermost Mobile), go to Permissions, Select "Grant admin consent for... " and follow the flow there, once completed attempt to login again on the mobile app again.

If any authentication path resolves a different identifier, enrollment will fail.

Supported Identity Attributes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Only the identity attributes listed below are supported for Intune MAM.

+-------------------+------------------+------------------------------+
| Attribute         | Supported        | Result                       |
+===================+==================+==============================+
| objectId          | Required         | Works                        |
+-------------------+------------------+------------------------------+
| email             | Not supported    | Enrollment fails             |
+-------------------+------------------+------------------------------+
| preferred_username| Not supported    | Identity mismatch            |
+-------------------+------------------+------------------------------+
| objectGUID        | Not supported    | Breaks mobile authentication |
+-------------------+------------------+------------------------------+
| Custom attributes | Not supported    | Unsupported by Intune        |
+-------------------+------------------+------------------------------+

> Attribute Synchronization & Access Enforcement

Reviewer: not sure if it should be mentioned here or somewhere else.

If Auth Provider is `SAML` users that have not been provisioned must do so first using the web or desktop app, failed to do this will result in an error message on the mobile app asking users to first login to their account using web or desktop.

Let me know if this point is not clear.

Author: What is the exact user-facing error string shown in the mobile app when this happens?

* User attributes synchronize only at login.
* Changes made in the identity provider do not apply until the next login.
* Mobile-only users may not trigger attribute synchronization.

As a result, attribute-based access control (ABAC) may not apply immediately.

If proactive enforcement of attribute-based access changes is required, we recommend LDAP (including Entra ID Domain Services). This behavior affects **access enforcement**, not Intune MAM enrollment.

Runtime Enforcement Behavior
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Mattermost Mobile enforces Intune MAM requirements during active sessions, not only at login.

If Intune MAM becomes newly required due to policy, licensing, or configuration changes:

* Enrollment is triggered immediately.
* Access to sensitive content is restricted until enrollment succeeds.
* Users cannot bypass enforcement.

Plan rollouts assuming enforcement can occur instantly.

Once your identity model and enforcement behavior are understood and aligned, ensure the following prerequisites are in place before beginning configuration.

Detailed Prerequisites
----------------------

Microsoft Requirements
~~~~~~~~~~~~~~~~~~~~~~~

* Microsoft Entra tenant

> * Permissions to register and manage enterprise applications

Reviewer: not sure how enterprise applications and App registrations differ, but they will need access to both

* Microsoft Intune App Protection Policies enabled
* Microsoft Entra–backed sign-in functions for web and mobile
* Targeted users are licensed for Microsoft Intune

Mattermost Requirements
~~~~~~~~~~~~~~~~~~~~~~~

* Mattermost Enterprise Advanced license

> * Microsoft Entra (OIDC) SSO configured

Reviewer: make sure this matches Step 2

* Mattermost mobile app registered as a public/native Entra application
* iOS platform configured with the correct bundle ID and redirect URI

> * ``accessTokenAcceptedVersion = 2`` set in the app manifest to ensure access tokens include required claims (such as ``oid``) in the correct format

Reviewer: we should mention this in Step 2 as well

* Use ``msDS-aadObjectId`` as the identity attribute.
* Do not use ``objectGUID``.

Mandatory Microsoft Entra API Permissions
-----------------------------------------

The following Microsoft Entra API permissions are mandatory for Intune MAM enrollment and server-side token validation. Missing permissions will block enrollment.

Grant admin consent for each required permission below:

+-----------------------+--------------------------------------------------------------+
| Component             | Permission                                                   |
+=======================+==============================================================+
| Mattermost Mobile App | Microsoft Mobile Application Management →                    |
|                       | ``user_impersonation`` (Delegated)                           |
+-----------------------+--------------------------------------------------------------+
| Mattermost Server     | ``https://msmamservice.api.application/.default``            |
+-----------------------+--------------------------------------------------------------+

.. note::

    ``user_impersonation`` appears as ``DeviceManagementManagedApps.ReadWrite``.

> With prerequisites met and identity requirements understood, proceed to the configuration steps in the next section.

Reviewer: This is us not the customers.

Required Mattermost Server Configuration
----------------------------------------

1. Go to **System Console > Environment > Mobile App Management**.
2. Enable **Microsoft Intune App Protection Policies**.
3. Set ``IdAttribute`` to ``objectId``.
4. Save your changes.

Validation Checklist
--------------------

Before rolling out to production, validate the configuration using a test user account. Confirm the following values match for the same user:

* Azure AD ``objectId``
* MSAL access token ``oid`` claim
* SAML ``objectidentifier`` (if applicable)
* LDAP ``msDS-aadObjectId`` (if applicable)

Any mismatch will cause Intune MAM enrollment to fail.

Deploy or Update Mattermost Mobile Apps
---------------------------------------

Install the latest Mattermost iOS mobile app using one of the following methods:

* Public App Store
* Microsoft Intune deployment
* Private or wrapped builds

.. note::

>    Mattermost Beta (``com.mattermost.rnbeta``) and Production (``com.mattermost.rn``) apps require **separate** Microsoft Entra app registrations.

Reviewer: this is not true either, both can be configured with the same Entra app in the Expose an API app registration

MDM device enrollment is not required.

Configure Intune App Protection Policies
----------------------------------------

1. Go to the **Microsoft Intune Admin Center**.
2. Create an **iOS App Protection Policy**.
3. Add the Mattermost bundle ID.
4. Assign the policy using **Azure AD groups**.

Mattermost roles do not control Intune MAM enforcement.

Expected Mobile Login & Enrollment Flow
---------------------------------------

When Intune MAM is enabled:

1. The user taps **Sign in with Microsoft**.
2. MSAL authenticates the user.
3. Mattermost validates the access token.
4. The mobile app checks:
   
   * Platform is iOS
   * Intune MAM is enabled
   * Authentication service matches
   * Required scope is populated
   * License is **Enterprise Advanced**

5. Intune MAM enrollment is triggered.

> 6. App protection policies are applied.

Reviewer: ```suggestion
1. The mobile app checks:
   * Platform is iOS
   * Intune MAM is enabled
   * Authentication service matches
   * License is **Enterprise Advanced**

2. The user taps **Sign in**. careful with using `Microsoft` explicitly cause we don't enforce it
3. MSAL authenticates the user.
4. Mattermost validates the access token.
5. Intune MAM enrollment is triggered.
6. App protection policies are applied.
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Mattermost Mobile App enforces Intune MAM requirements during active sessions, not only at login.

If Intune MAM becomes newly required due to policy, licensing, or configuration changes:

* Enrollment is triggered immediately.
* Access to sensitive content is restricted until enrollment succeeds.
* Users can't bypass enforcement.

Plan rollouts assuming enforcement can occur instantly.

Once your identity model and enforcement behavior are understood and aligned, ensure the following prerequisites are in place before beginning configuration.

Microsoft Entra Configuration for Intune MAM
--------------------------------------------

This section provides the detailed Microsoft Entra configuration required to support Mattermost Mobile App authentication and Intune MAM enforcement. Complete this section before 
`configuring the Mattermost server <#configure-mattermost-server>`__ or `Intune App Protection Policies <#configure-intune-app-protection-policies>`__.

The steps below require changes in App registrations (manifest + API permissions) and Enterprise applications (admin consent).

Entra Application Registration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Register an application in Microsoft Entra that will be used by the Mattermost Mobile App for authentication and Intune MAM enrollment. This application represents the Mattermost Mobile client and is used to acquire MSAL access tokens during mobile sign-in.

Reviewer: > Register an application in Microsoft Entra that will be used by the Mattermost Mobile App for authentication and Intune MAM enrollment. This application represents the Mattermost Mobile client and is used to acquire MSAL access tokens during mobile sign-in.

Register an application in Microsoft Entra that will be used by the Mattermost Mobile App for authentication and Intune MAM enrollment. This application ~~represents~~ is used by the server to resolve the claims of the access token acquired by MSAL during mobile sign-in.

* ``SamlSettings.IdAttribute`` must map to ``objectidentifier``.
* Email, UPN, and ``immutableID`` are not supported.

.. important::

  When SAML is selected as the authentication method for Intune MAM enforcement, users must already exist in Mattermost before signing in on mobile. Users who haven't yet been provisioned must first sign in using the Mattermost web or desktop application. Mobile sign-in doesn't create new users for SAML-based authentication. If a user attempts to sign in on mobile before being provisioned, the user will be prompted to sign in using web or desktop.

LDAP (Entra ID Domain Services)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Use ``msDS-aadObjectId`` as the identity attribute.
* Do not use ``objectGUID``.

With prerequisites met and identity requirements understood, proceed to the configuration steps in the next section.

Configure Mattermost Server
----------------------------

1. Go to **System Console > Environment > Mobile Security**.
2. Enable **Microsoft Intune App Protection Policies**.
3. Configure the following fields using values from the Microsoft Entra application created earlier:
   
   * **Application (Client) ID**
   * **Directory (Tenant) ID**

>    * **Authentication Provider**:

Reviewer: this goes first

     * OIDC (Microsoft Entra-backed), or
     * SAML (backed by Microsoft Entra)

> 4. Set ``IdAttribute`` to ``objectId``.

Reviewer: this is not relevant in that section

* Mattermost Enterprise Advanced licensing is available for the server.
* Target users are licensed for Microsoft Intune.
* You can register applications and grant tenant-wide admin consent in Microsoft Entra.
* If enforcing Intune MAM on SAML, users already exist in Mattermost (mobile sign-in does not create users for SAML).

.. note::

  In this guide, OpenID Connect (OIDC) refers to the Microsoft Entra sign-in method used by the Mattermost Mobile App via MSAL.

Configuration Overview
----------------------

Successful Intune MAM enforcement requires coordinated configuration across:

* **Microsoft Entra ID** – app registration, scope, API permissions, admin consent
* **Microsoft Intune** – iOS App Protection Policies
* **Mattermost Server** – MAM enablement and provider selection
* **Mattermost Mobile App (iOS)** – enrollment and enforcement

Complete the steps below in order.

Setup Summary
-------------

1. Confirm identity requirements and alignment (``objectId ↔ oid``).
2. Configure Microsoft Entra (server-referenced application).
3. Enable Intune MAM in Mattermost and select the enforced provider.
4. Deploy the official Mattermost iOS app.

> 5. Configure Intune App Protection Policies.

Reviewer: I would have this as number 3.

1. Make sure your use ``objectId ↔ oid`` in your current Mattermost authentication method
2.  Configure Entra (server-referenced application)
3. Configure policies on Intune
4. Enable Intune MAM in Mattermost and select the enforced provider that matches step 1
5. Download and install the app
6. Validate enrollment and enforcement

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Authorize the official Mattermost Mobile client application ID (provided by Mattermost) to request the ``login.mattermost`` scope.

1. In the Entra app registration, go to **Expose an API**.
2. Under **Authorized client applications**, select **Add a client application**.
3. Add the official Mattermost Mobile client application ID (provided by Mattermost).
4. Authorize the ``api://<APPLICATION-ID>/login.mattermost`` scope.
5. Save your changes.

If the official client application ID isn't yet available, complete the remaining steps and return here once Mattermost provides it.

API Permissions and Admin Consent
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. _api-permissions-admin-consent:

1. Go to **API permissions**.
2. Add the required Intune MAM permissions for enrollment (minimum required):

   * ``https://msmamservice.api.application/.default``
   * Microsoft Mobile Application Management → ``user_impersonation`` (Delegated)

3. Grant **tenant-wide admin consent**.

If these permissions are missing or lack tenant-wide admin consent, enrollment can fail with an Entra permissions error (for example, ``AADSTS650057``).

> Configure MSAL v2 Tokens

Reviewer: Before this step

Go to Token Configuration
* Add optional claims
* Select "Access" for the token type
* Add the claims
    - email
    - family_name
    - given_name
    - preferred_username
    - upn

          <span class="label">Next step:</span> Verify Enterprise Advanced licensing, Intune MAM is enabled, <strong>Auth Provider</strong> selection matches how users authenticate, configured <strong>Tenant ID</strong> and <strong>Application (Client) ID</strong> are correct, and tenant-wide admin consent is granted. Then confirm identity alignment (<code>objectId ↔ oid</code>) and Intune App Protection Policy targeting for the correct iOS bundle ID.
        </td>
      </tr>

    </tbody>
  </table>

Consent During First Login
~~~~~~~~~~~~~~~~~~~~~~~~~~~

If a user’s first mobile sign-in fails with a message such as **Consent Denied** or: ``You denied consent for Intune management. The affected accounts have been unenrolled and signed out.``

Treat this as a tenant-wide admin consent issue for the Intune MAM permissions on the Entra app registration configured in Mattermost Server (Step 3). The message may appear as if the user denied consent, but the underlying issue is that required admin consent has not been granted for the tenant.

To resolve this:

> 1. In Microsoft Entra, grant **tenant-wide admin consent** for the required Intune MAM permissions on the same Entra app registration configured in Mattermost.

Reviewer: see the above comment, this is in Enterprise apps

You don't register the Mattermost Mobile app itself. Redirect URI configuration isn't required.

Entra Application Registration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Go to **Identity > Applications > App registrations**
2. Select **New registration**
3. Configure:

   * **Name**: Mattermost Mobile (Intune MAM)
   * **Supported account types**: Single tenant

4. Register the app
5. Copy:

   * **Application (client) ID**
   * **Directory (tenant) ID**

Expose the API and Create the Scope
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Go to **Expose an API**
2. Confirm **Application ID URI** is set (for example, ``api://<APPLICATION-ID>``)
3. Add a scope named ``login.mattermost``
4. Save your changes.

Authorize the Official Mobile Client
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Authorize the official Mattermost Mobile client application ID (provided by Mattermost) to request the ``login.mattermost`` scope.

1. In the Entra app registration, go to **Expose an API**.
2. Under **Authorized client applications**, select **Add a client application**.

> 3. Add the official Mattermost Mobile client application ID (provided by Mattermost) as ``64e9952b-20eb-46dc-92ad-99089ed24903``.

Reviewer: ```suggestion
3. Add the official Mattermost Beta Mobile client application ID (provided by Mattermost) as ``64e9952b-20eb-46dc-92ad-99089ed24903``.
```

> 3. Add the official Mattermost Beta Mobile client application ID (provided by Mattermost) as ``64e9952b-20eb-46dc-92ad-99089ed24903``.

Reviewer: Beta is correct and the clientId matches. Once @iyampaul adds the one for production in Entra, then we will add that one too
